test/Spec.hs

```diff
@@ -46,11 +46,19 @@ tests docker0 = testGroup "Tests"
       (expectFailure (withMaxSuccess 2000 prop_dieHard))
   , testGroup "MemoryReference"
       [ testProperty "NoBug"                            (prop_sequential None)
-      , testProperty "LogicBug"          (expectFailure (prop_sequential Logic))
-      , testProperty "RaceBugSequential"                (prop_sequential Race)
-      , testProperty "RaceBugParallel"   (expectFailure (prop_parallel   Race))
-      , testProperty "PreconditionFailed" prop_precondition
-      , testProperty "ExistsCommands"     prop_existsCommands
+       , testProperty "LogicBug"          (expectFailure (prop_sequential Logic))
+       , testProperty "RaceBugSequential"                (prop_sequential Race)
+       , testProperty "RaceBugParallel"   (expectFailure (prop_parallel   Race))
+       , testProperty "PreconditionFailed" prop_precondition
+       , testProperty "ExistsCommands"     prop_existsCommands
+       , testProperty "NoBug 1 thread"           (prop_general_parallel None 1)
```

Comment:
Perhaps you can check that `prop_general_parallel bug 1` is equivalent to `prop_sequential bug` and `prop_general_parallel bug 2` is equivalent `prop_parallel bug`? In the parallel case it might not be so easy because of the non-determinism of the scheduler, but perhaps you could check that the shrinking part at least matches up, similar to how it's done in: https://github.com/advancedtelematic/quickcheck-state-machine/blob/master/test/ShrinkingProps.hs .

Reply:
I don't think shrinking matches up atm. The difference is that when you shrink a pair of commands it always consists of 2 commands. However when you shrink a list of 2 it may actually result in a singleton or an empty list. I think I should change the shrinking so that it retains the length of the list and it rather only shrinks the elements. This will also give a better fix on the bug that appeared.

Reply:
For example `shrinkPairS' shrinkListS' ([1,2],[3,4])` gives
```
[Shrunk {wasShrunk = True, shrunk = ([],[3,4])},Shrunk {wasShrunk = True, shrunk = ([2],[3,4])},
Shrunk {wasShrunk = True, shrunk = ([1],[3,4])},Shrunk {wasShrunk = True, shrunk = ([1,2],[])},
Shrunk {wasShrunk = True, shrunk = ([1,2],[4])},Shrunk {wasShrunk = True, shrunk = ([1,2][3])}, Shrunk {wasShrunk = False, shrunk = ([1,2],[3,4])}]
```
but `shrinkListS shrinkListS' [[1,2],[3,4]]` gives
```
[Shrunk {wasShrunk = True, shrunk = []},Shrunk {wasShrunk = True, shrunk = [[3,4]]},
Shrunk {wasShrunk = True, shrunk = [[1,2]]},
Shrunk {wasShrunk = True, shrunk = [[][3,4]]},Shrunk {wasShrunk = True, shrunk = [[2],[3,4]]},
Shrunk {wasShrunk = True, shrunk = [[1] [3,4]]},Shrunk {wasShrunk = True, shrunk = [[1,2],[]]},
Shrunk {wasShrunk = True, shrunk = [[1,2],[4]]},Shrunk {wasShrunk = True, shrunk = [[1,2],[3]]},
Shrunk {wasShrunk = False, shrunk = [[1,2],[3,4]]}]
```
You can see that the first three elements should be fitlered out, so that there is some sence of equivalence. Doing this filter makes shrinking less agressive, but maybe it is good as a first step to check the equivalence with the older implementation.

Reply:
Hmm, sometimes too aggressive shrinking isn't good either. For example, if either of the two inner lists are empty then we are effectively executing the commands on a single thread.